File: desktop/src/lib/launchOptions.ts
// Mirrors monadeck-core's launch_options::steam_launch_options. Kept client-side
// so the string updates live as Settings change (no IPC round-trip / race).
import type { MonadeckConfig } from "./types";

/// The virtual pad's USB id (mirrors core::vr_games::PAD_DEVICE).
const PAD_DEVICE = "0x045e/0x028e";

function quote(v: string): string {
  return /\s/.test(v) ? `"${v}"` : v;
}

export function steamLaunchOptions(cfg: MonadeckConfig | null): string {
  if (!cfg) return "";
  // Client-side only: env vars in config.environment go to monado-service, NOT
  // the game (Envision puts only the pressure-vessel flag here too).
  const parts: string[] = [];
  if (cfg.ovr_runtime === "xrizer" && cfg.xrizer_path) {
    parts.push(`VR_OVERRIDE=${quote(cfg.xrizer_path)}`);
  }
  parts.push("PRESSURE_VESSEL_IMPORT_OPENXR_1_RUNTIMES=1");
  // Keep gaming mode's virtual Xbox pad out of VR games (they would read it as
  // their own gamepad). GE-style Protons get this from a local fix the overlay
  // writes; Valve's Proton only honours launch options.
  parts.push(`SDL_GAMECONTROLLER_IGNORE_DEVICES=${PAD_DEVICE}`);
  parts.push(`SDL_JOYSTICK_BLACKLIST_DEVICES=${PAD_DEVICE}`);
  parts.push("%command%");
  return parts.join(" ");
}

File: crates/core/src/launch_options.rs
//! Builds the Steam launch-options string games need to pick up monado + xrizer.
//!
//! This is where Monadeck improves on Envision: Envision's helper only emits
//! `PRESSURE_VESSEL_IMPORT_OPENXR_1_RUNTIMES=1 %command%` and relies on
//! `openvrpaths.vrpath` for OpenVR discovery — but that registration doesn't
//! survive the Proton / Steam-Linux-Runtime sandbox, so OpenVR games need an
//! explicit `VR_OVERRIDE` pointing straight at the xrizer runtime. We include it.

use crate::config::{MonadeckConfig, OvrRuntime};
use crate::paths::home;
use std::path::PathBuf;

/// Wrap in double quotes only when the value would otherwise split on whitespace.
fn quote(value: &str) -> String {
    if value.chars().any(char::is_whitespace) {
        format!("\"{value}\"")
    } else {
        value.to_string()
    }
}

/// The launch-options line, ready to paste into a Steam game's properties.
///
/// Client-side only: `VR_OVERRIDE` (when registering xrizer) + the pressure-
/// vessel import flag + `%command%`. The user's `config.environment` vars are
/// server-side (they go to monado-service), so they are deliberately NOT
/// included here.
pub fn steam_launch_options(cfg: &MonadeckConfig) -> String {
    let mut parts: Vec<String> = Vec::new();

    if cfg.ovr_runtime == OvrRuntime::Xrizer {
        if let Some(xr) = cfg.xrizer_path.as_ref() {
            parts.push(format!("VR_OVERRIDE={}", quote(&xr.to_string_lossy())));
        }
    }

    parts.push("PRESSURE_VESSEL_IMPORT_OPENXR_1_RUNTIMES=1".to_string());
    parts.push("%command%".to_string());
    parts.join(" ")
}

/// Best-effort guess at the xrizer runtime dir, following the layout recommended
/// on the xrizer README: `~/.local/share/xrizer/<version>`, preferring
/// `xrizer-nightly`, else the first version directory present.
pub fn detect_xrizer_path() -> Option<PathBuf> {
    let base = home().join(".local/share/xrizer");
    let nightly = base.join("xrizer-nightly");
    if nightly.is_dir() {
        return Some(nightly);
    }
    let mut dirs: Vec<PathBuf> = std::fs::read_dir(&base)
        .ok()?
        .flatten()
        .map(|e| e.path())
        .filter(|p| p.is_dir())
        .collect();
    dirs.sort();
    dirs.into_iter().next_back()
}
